presto-hive/src/main/java/com/facebook/presto/hive/pagefile/PageFileFooterReader.java

```
@@ -0,0 +1,119 @@
+
+    private final List<Long> stripeOffsets;
+    private final long footerOffset;
+    private final Optional<HiveCompressionCodec> compressionCodec;
+
+    public PageFileFooterReader(
+            FSDataInputStream inputStream,
+            long fileSize)
+            throws IOException
+    {
+        checkArgument(fileSize >= FOOTER_LENGTH_IN_BYTES, "Malformed PageFile format, footer length is missing.");
+        requireNonNull(inputStream, "inputStream is null");
+        ImmutableList.Builder<Long> stripeOffsetsBuilder = ImmutableList.builder();
+
+        byte[] buffer = new byte[toIntExact(min(fileSize, ESTIMATED_FOOTER_SIZE))];
+        inputStream.readFully(fileSize - buffer.length, buffer);
+        int footerSize = Slices.wrappedBuffer(buffer, buffer.length - FOOTER_LENGTH_IN_BYTES, FOOTER_LENGTH_IN_BYTES).getInt(0);
+
+        footerOffset = fileSize - footerSize;
+        HiveCompressionCodec compression;
+        if (footerOffset < 0) {
+            throw new PrestoException(HIVE_BAD_DATA, "Malformed PageFile format, incorrect footer length.");
+        }
+        else if (footerOffset > 0) {
+            if (footerSize > buffer.length) {
+                buffer = new byte[footerSize];
+                inputStream.readFully(footerOffset, buffer);
+            }
+
+            FixedLengthSliceInput sliceInput = Slices.wrappedBuffer(buffer, buffer.length - footerSize, footerSize - FOOTER_LENGTH_IN_BYTES).getInput();
+            long remainingSize = sliceInput.length();
+            // read compression
+            int compressionStringSize = sliceInput.readInt();
+            remainingSize -= SIZE_OF_INT;
+            String compressionName = sliceInput.readSlice(compressionStringSize).toStringUtf8();
+            remainingSize -= compressionStringSize;
+
+            try {
+                compression = HiveCompressionCodec.valueOf(compressionName);
+            }
+            catch (Exception e) {
```

Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)